src/cpp/rtps/messages/RTPSMessageGroup.cpp

```diff
@@ -477,19 +573,30 @@ bool RTPSMessageGroup::add_data(
     InlineQosWriter* inline_qos;
     inline_qos = (change.inline_qos.length > 0 && nullptr != change.inline_qos.data) ? &qos_writer : nullptr;
 
+    bool copy_data = false;
 #if HAVE_SECURITY
     uint32_t from_buffer_position = submessage_msg_->pos;
+    bool protect_payload = endpoint_->getAttributes().security_attributes().is_payload_protected;
+    bool protect_submessage = endpoint_->getAttributes().security_attributes().is_submessage_protected;
+    bool protect_rtps = participant_->security_attributes().is_rtps_protected;
+    copy_data = protect_payload || protect_submessage || protect_rtps;
 #endif // if HAVE_SECURITY
     const EntityId_t& readerId = get_entity_id(sender_->remote_guids());
 
+    // If gather-send is possible, get payload
+    if (!copy_data)
+    {
+        get_payload(change);
+    }
+
```

Comment:
Only do this when everything in the method went well (just before returning true). We don't want to keep a reference to a payload if it is not going to be sent.